### ISSUE TITLE
chromedp.WaitVisible("#foot") hangs but WaitVisible("foot", chromedp.ByID) works

### ISSUE THREAD

**Issue by kjk** [chromedp/chromedp#440]
Go: `go version go1.12.7 windows/amd64`

chromedp:
```
github.com/chromedp/cdproto v0.0.0-20190721111337-61a0348ea0b1
github.com/chromedp/chromedp v0.3.1
```

Full test program: https://gist.github.com/kjk/804bca2d2dfc102f73d8817d4997d172

The core:
```go
qe := url.QueryEscape(searchTerm)
uri := fmt.Sprintf("https://www.google.com/search?q=" + qe)
fmt.Printf("before run, uri='%s'\n", uri)
sel := "div.rc > div:first-child > a:first-child"
var nodes []*cdp.Node
err := chromedp.Run(ctx,
	chromedp.Navigate(uri),
	chromedp.WaitVisible("#foot"),
	chromedp.Nodes(sel, &nodes),
)
```
This hangs in `WaitVisible("#foot").
It works when I replace:
```go
chromedp.WaitVisible("#foot"),
```
with:
```go
chromedp.WaitVisible(`foot`, chromedp.ByID),
```

Those should be equivalent but it looks like `BySearch` version doesn't work.

I also verified in dev tools console that `$("#foot")` works (which I assume uses the same browser API as `WaitVisible("#foot")` would end up using).

**Comment by mvdan** [comment]
Thanks for the detailed report. Did you test with master by any chance?

**Comment by kjk** [comment]
As far as I can tell, it also happens on master.

**Comment by mvdan** [comment]
Thanks. I typed the following reply three days ago, but I think it got lost in the ether due to the poor connection I had at the time.

The misconception here is that the default `By` func isn't `ByQuery`, but `BySearch` - which is pretty much a plaintext search over the page's HTML. And because the `#foo` string appears three times, and not all of those occurrences are contained by visible nodes, the `Wait` function hangs forever.

Try `chromedp.WaitVisible("#foot", chromedp.ByQuery)` - I'm willing to bet it will work, because `ByID` is simply that, plus ensuring the leading `#`.

The default is documented, as you can see on the godoc:

>  By* options specify the type of element query used By the browser to perform the selection query. When not specified, element queries will use BySearch (a wrapper for DOM.performSearch). 

However, that perhaps isn't very obvious or clear. If you have a suggestion to improve the docs or add an entry to the FAQ, I'll gladly accept a PR.

**Comment by kjk** [comment]
> The misconception here is that the default By func isn't ByQuery, but BySearch - which is pretty much a plaintext search over the page's HTML. And because the #foo string appears three times, and not all of those occurrences are contained by visible nodes, the Wait function hangs forever.

To paraphrase the explanation:

> WaitVisible BySearch uses DOM.performSearch("#foot") and DOM.performSearch("#foot") doesn't return anything because there are visible and non-visible matches for the "#foot" query

I don't know how to easily prove / disapprove this or to verify that this is how `WaitVisible()` is implemented.

But the explanation doesn't match my expectations and I don't see a way to document it away.

There is a visible node with id `#foot` so `WaitVisible("#foot")` should work.

The docs for `performSearch` https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-performSearch say:
> query : Plain text or query selector or XPath search query.

I don't see an interpretation that would result in "doesn't return any results".

If `#foot` is interpreted as query selector (most likely), then it should match the visible node with that id.

If it's interpreted as plain text then it should return at least one result. The doc doesn't say how plain text search deals with invisible nodes, but it should at least return the result for visible nodes.

For what it's worth, the following pupeteer script works in both cases:

```javascript
const puppeteer = require('puppeteer');

const uri = 'https://google.com/search?q=gist';
const sel = "#foot";

async function do1() {
    const browser = await puppeteer.launch();
    const page = await browser.newPage();
    await page.goto(uri);
    await page.waitFor(sel, { visible: true });
    await browser.close();
}

async function do2() {
    const browser = await puppeteer.launch();
    const page = await browser.newPage();
    await page.goto(uri);
    await page.waitForSelector(sel, { visible: true });
    await browser.close();
}

(async () => {
    console.log("before do1");
    await do1();
    console.log("before do2");
    await do2();
})()
```

I assume that `do1()`, using `page.waitFor("#foot", {visible: true})`, should be equivalent to `WaitVisible("#foot")`

**Comment by mvdan** [comment]
> I don't see an interpretation that would result in "doesn't return any results".

Sorry, that's not what I meant. Apologies if my comment was unclear.

The plaintext search returns three results; the same one you get via `ByQuery`, and two others. Either of the two others belongs to a non-visible HTML element, hence `WaitVisible` hangs forever. You can see the three matches if you view the raw HTML and use Control+F, or at least I did when I tried a few days ago.

`performSearch` is a useful default, because it often does what the user wants, even if it's not a query selector. It's only in edge cases like this one that it might do something unexpected. In that case, you should use `ByQuery` or `ByID`. In other words, if you know you're using a query, `ByQuery` is going to be a narrower search and thus be more stable.

This is what I'm suggesting to amend in the godoc or FAQ. I can see myself being confused in a similar way.

I'm of course open to more suggested fixes if you have ideas. Changing the default is always an option, but I'd rather not break existing programs.

**Comment by rhcarvalho** [comment]
> `performSearch` is a useful default, because it often does what the user wants, even if it's not a query selector. It's only in edge cases like this one that it might do something unexpected. In that case, you should use `ByQuery` or `ByID`. In other words, if you know you're using a query, `ByQuery` is going to be a narrower search and thus be more stable.


Took me quite some time to realize I was hitting the "trap" that `performSearch` also matches plain text.

An action like `chromedp.SendKeys("#myid", "123")` was hanging because `#myid` appears in the middle of some JavaScript code in the page, and is matched as a `Text` node.

That was rather surprising behavior.

Basically if the selector is literally found in the page, be it in JavaScript code, a `<p>`, an HTML comment, anywhere, then suddenly `#myid` is not matching a single element anymore.


> This is what I'm suggesting to amend in the godoc or FAQ. I can see myself being confused in a similar way.

[`BySearch`](https://godoc.org/github.com/chromedp/chromedp#BySearch) today mentions CSS selectors and XPath, not freeform text.

![image](https://user-images.githubusercontent.com/88819/77760293-613aa680-7036-11ea-9618-be6bf4475c28.png)

The [devtools protocol](https://github.com/ChromeDevTools/devtools-protocol/blob/3a31fb959097a363b3dd0941f18cc917305c2ce8/pdl/browser_protocol.pdl#L1918-L1930) says "Plain text or query selector or XPath search query."

I can send a PR to update the godoc.

**Comment by rhcarvalho** [comment]
(my opportunity to say thanks for `chromedp` to all who've worked on it, I love it! Used it last year, and found a use for it again now)

> `performSearch` is a useful default, because it often does what the user wants, even if it's not a query selector

I see it would be useful to have a way to say "wait until 'this sentence' is visible", without interpreting `'this sentence'` as neither CSS selector nor XPath query. In the cases where one specifically writes a CSS selector it is surprising that it matches arbitrary text in the HTML (visible or not).

Is there an advantage to `BySearch` over three explicit modes `ByText`, `ByCSS`, `ByXPath`?

**Comment by mvdan** [comment]
I agree that mixing conceptually different types of queries is a bit weird, but that's how the Chrome devtools protocol works. See https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-performSearch:

> query    string
>    Plain text or query selector or XPath search query.

**Comment by gervasiomatt** [comment]
Thanks for all the insight in this issue. I also ran into the same problem and after spending over a day on it discovered this thread. It wasn't obvious to me the difference between `BySearch` and ByQuery` and I didn't see it called out in the docs for `WaitVisible`

**Comment by ZekeLu** [comment]
#590 is merged. I think it's okay to close this issue now.
